include/ginkgo/core/base/precision_dispatch.hpp

```diff
@@ -68,29 +68,21 @@ namespace gko {
  * @tparam ValueType  the value type into whose associated matrix::Dense type to
  *                    convert the input LinOp.
  */
-template <typename ValueType>
-detail::temporary_conversion<matrix::Dense<ValueType>>
-make_temporary_conversion(LinOp* matrix)
-{
-    auto result =
-        detail::temporary_conversion<matrix::Dense<ValueType>>::template create<
-            matrix::Dense<next_precision<ValueType>>>(matrix);
-    if (!result) {
-        GKO_NOT_SUPPORTED(matrix);
-    }
-    return result;
-}
-
-
-/** @copydoc make_temporary_conversion(LinOp*) */
-template <typename ValueType>
-detail::temporary_conversion<const matrix::Dense<ValueType>>
-make_temporary_conversion(const LinOp* matrix)
+template <typename ValueType, typename Ptr>
+detail::temporary_conversion<std::conditional_t<
+    std::is_const<detail::pointee<std::remove_reference_t<Ptr>>>::value,
+    const matrix::Dense<ValueType>, matrix::Dense<ValueType>>>
```

Comment:
we could extract that into a helper struct to keep this and the one below more concise.